firedrake/adjoint/ensemble_reduced_functional.py

```diff
@@ -53,15 +61,72 @@ class EnsembleReducedFunctional(ReducedFunctional):
     works, please refer to the `Firedrake manual
     <https://www.firedrakeproject.org/parallelism.html#id8>`_.
     """
-    def __init__(self, J, control, ensemble, scatter_control=True):
-        super(EnsembleReducedFunctional, self).__init__(J, control)
+    def __init__(self, J, control, ensemble, scatter_control=True,
+                 gather_functional=None):
         self.ensemble = ensemble
+        if isinstance(J, list):
+            assert isinstance(control, list), "Controls should also be a list."
+            assert len(control) % len(J) == 0, "Controls and J have mismatching lengths."
+            self.controls_per_J = len(control) // len(J)
+            self.functional = J
+            self.controls = Enlist(control)
+            self.Jhats = []
+            self.sizes = self.ensemble.ensemble_comm.allgather(len(J))
+            for i in range(len(J)):
+                i0 = self.controls_per_J*i
+                i1 = self.controls_per_J*(i+1)
+                self.Jhats.append(ReducedFunctional(J[i], control[i0:i1]))
+        else:
+            super(EnsembleReducedFunctional, self).__init__(J, control)
```

Comment:
don't have this case. Just treat this as a list of one `J`.

Reply:
That will break the optimisation wrappers.

Reply:
That can't really be true because you don't call the superclass constructor in the case where you have a list of Js.

Reply:
The optimisation wrappers already don't work in that case. I'm concerned about breaking it for the usual case.

Reply:
I think I am confused about what you are asking me to do. Are you asking me to always call the superclass constructor? I cannot do that when J is a list.

Reply:
No. I'm saying never call the superclass constructor. In fact if we were really doing this properly then we wouldn't actually subclass ReducedFunctional but we'd instead make an AbstractReducedFunctional and inherit from that.

The optimisation wrappers will need to be fixed anyway, right?

Reply:
I'm not proposing to touch the optimisation wrappers. ROL is a binfire and the allgather in the scipy wrappers is not feasible for ensemble in this case.

Reply:
So what are you going to use a ReducedFunctional for if not to optimise?

Reply:
Initially, compute a gradient. Later, I will have to hand tool my own Gauss-Newton solver because none of the packages provide it, and all other algorithms are too slow.

Reply:
Eventually the TAO situation will be sorted out and then I can fix wrappers for that.

Reply:
Actually having read the code more carefully, I see that it can work just fine if we always Enlist J and make Jhats. I'll sketch it out in the following commits.

Reply:
OK I have fixed all this now, I will add something to the PR description.